src/MongoDB.Driver.Core/Core/Authentication/SaslAuthenticator.cs

```diff
@@ -65,12 +65,45 @@ public void Authenticate(IConnection connection, ConnectionDescription descripti
             Ensure.IsNotNull(connection, nameof(connection));
             Ensure.IsNotNull(description, nameof(description));
 
+            var firstStepResult = description.IsMasterResult.SpeculativeAuthenticate;
+
+            if (firstStepResult == null)
+            {
+                Authenticate(connection, description, previousStep: null, previousStepResult: null, cancellationToken);
+            }
+            else
+            {
+                (var firstStep, _) = CreateFirstStepAndCommand(_mechanism, connection, conversation: null, description);
+                Authenticate(
+                    connection,
+                    description,
+                    previousStep: firstStep,
+                    previousStepResult: firstStepResult,
+                    cancellationToken);
+            }
+        }
+
+        private void Authenticate(
```

Comment:
An alternative name for this could be "ResumeAuthenticate"

Reply:
My initial design allowed Authentication to be continued at **ANY** step, which is probably overkill for this PR. 
So I agree with @rstam  and @DmitryLukyanov  that resulted in an unneeded overload that was actually probably should have been a separate method named "ResumeAuthenticate".

@rstam : I do believe your refactor is correct, however, I find the need to check if command is null every single iteration of the loop to be somewhat confusing (especially since such a condition should really only happen once. 

@rstam @DmitryLukyanov : I've refactored my PR and pushed the changes: What do you think of this new refactored version? I think the key feature of my refactor is that the loop feels simpler to me: all the special case logic happens prior to the loop starting: the loop simply executes the current command and creates the next command.

Reply:
Since this thread is "oudated", I'm resolving this thread in favor of the new thread.